mono/utils/checked-build.c

```diff
@@ -225,4 +229,279 @@ assert_gc_neutral_mode (void)
+	g_assertion_message("Mempool reference violation: " __VA_ARGS__)
+
+// Say image X "references" image Y if X either contains Y in its modules field, or X’s "references" field contains an
+// assembly whose image is Y.
+// Say image X transitively references image Y if there is any chain of images-referencing-images which leads from X to Y.
+// Once the mempools for two pointers have been looked up, there are four possibilities:
+
+// Case 1. Image FROM points to Image TO: Legal if FROM transitively references TO
+
+// We'll do a simple BFS graph search on images. For each image we visit:
+static void
+check_image_search (GHashTable *visited, GPtrArray *next, MonoImage *candidate, MonoImage *goal, gboolean *success)
+{
+	// Image hasn't even been loaded-- ignore it
+	if (!candidate)
+		return;
+
+	// Image has already been visited-- ignore it
+	if (g_hash_table_lookup_extended (visited, candidate, NULL, NULL))
+		return;
+
+	// Image is the target-- mark success
+	if (candidate == goal)
+	{
+		*success = TRUE;
+		return;
+	}
+
+	// Unvisited image, queue it to have its children visited
+	g_hash_table_insert (visited, candidate, NULL);
+	g_ptr_array_add (next, candidate);
+	return;
+}
+
+static gboolean
+check_image_may_reference_image(MonoImage *from, MonoImage *to)
+{
+	g_warning ("QQPTX: IMAGE->IMAGE %p (%s) seeks %p (%s)", from, from->name, to, to->name); // FIXME: Remove all QQPTX lines
+	if (to == from) // Shortcut
+		return TRUE;
+
+	// Corlib is never unloaded, and all images implicitly reference it.
+	// Some images avoid explicitly referencing it as an optimization, so special-case it here.
+	if (to == mono_defaults.corlib)
+		return TRUE;
+
+	gboolean success = FALSE;
+
+	// Images to inspect on this pass, images to inspect on the next pass
+	GPtrArray *current = g_ptr_array_sized_new (1), *next = g_ptr_array_new ();
+
+	// Because in practice the image graph contains cycles, we must track which images we've visited
+	GHashTable *visited = g_hash_table_new (NULL, NULL);
+
+	#define CHECK_IMAGE_VISIT(i) check_image_search (visited, next, (i), to, &success)
+
+	CHECK_IMAGE_VISIT (from); // Initially "next" contains only from node
+
+	// For each pass exhaust the "to check" queue while filling up the "check next" queue
+	while (!success && next->len > 0) // Halt on success or when out of nodes to process
+	{
+		// Swap "current" and "next" and clear next
+		GPtrArray *temp = current;
+		current = next;
+		next = temp;
+		g_ptr_array_set_size (next, 0);
+
+		int current_idx;
+		for(current_idx = 0; current_idx < current->len; current_idx++)
+		{
+			MonoImage *checking = g_ptr_array_index (current, current_idx); // CAST?
+
+			mono_image_lock (checking);
+
+			g_warning ("QQPTX: idx %d this pass, image %p (%s) has %d modules %d references (%p)", (int)current_idx, checking, checking->name, (int)checking->module_count, (int)checking->nreferences, checking->references); // FIXME: Remove all QQPTX lines
```

Comment:
Don't remove the log messages, you'll miss them when troubleshooting issues.

One pattern we have in the runtime is to hide them under a macro and disable by default.